### DISCUSSION TITLE
Go Serverless function in SvelteKit project

### DISCUSSION THREAD

**Discussion by beeb** [vercel/vercel#8885] [Help] [answered]
I have a sveltekit project that could use a golang serverless function for some heavier processing.

If I try to create an `api` folder at the root of my project, with a simple `hello.go` file in there, and provide the required `go.mod` file in the root, then I can't call this function at `domain.com/api/hello` since the sveltekit router takes precedence and tries to look for `src/routes/api/hello`.

Is there a way for me to use Go inside my sveltekit project or should I deploy a separate project for it?

Best,
Val

**Comment by amyegan** [comment, marked as answer]
Hi Val! I'm guessing you get a 404 at `domain.com/api/hello`, but please let me know if you see a different error or any other error details in [function logs](https://vercel.com/docs/concepts/deployments/logs#function-logs). Do you have a public repo or [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can use to get a better look at what is happening?

Without seeing the error details, I only have a guess. You may need to [configure the golang function](https://vercel.com/docs/project-configuration#project-configuration/functions) in your `vercel.json` project config to use the [Go Runtime](https://vercel.com/docs/concepts/functions/serverless-functions/runtimes/go#advanced-go-usage). Go is typically zero config, but I suspect the default Node.js Runtime is being used for the whole app since the build is detecting SvelteKit.

It might be easier to deploy as a separate project and use [rewrites](https://vercel.com/docs/project-configuration#project-configuration/rewrites) to make the URL match.

**Reply by beeb** [reply]
I tried to create a repro and the function was working as intended. This must be an issue with the fact that my main repo is a turborepo. Let me try to reproduce with a turborepo and I'll post here again.

EDIT: i didn't check if sveltekit endpoints were working during that test.

**Reply by beeb** [reply]
Hey there! So I managed to reproduce something similar. This time however, the sveltekit endpoint does not work anymore, and the go endpoint works.

https://github.com/beeb/repro-sveltekit-go
https://repro-sveltekit-go.vercel.app/api/hello (go endpoint)
https://repro-sveltekit-go.vercel.app/api/test (sveltekit endpoint)

It seems now the deployment is automatically configured to support Go, but there is an obvious conflict with the routes at `/api` since those should be routed by sveltekit when they exist, and only try Go if sveltekit reports a 404.

**Reply by amyegan** [reply]
🤔 I don't know enough about SvelteKit to identify what is happening, but this certainly feels like very strange behavior! I reached out to the Svelte experts to get their advice and will share an update here when I learn more.

**Reply by amyegan** [reply]
I'm sorry it took so long to get back with you! You probably already found your own way around the problem at this point, but I'm sharing the little bit that I learned in case it's helpful to anyone. The engineers' recommended solution was to avoid having any `src/routes/api` routes at all, and rely entirely on root level `/api` routes for serverless functions. Vercel should pick up all `/api` routes with this configuration so they should never make it to SvelteKit for routing.

**Reply by beeb** [reply]
Hey @amyegan . Thanks for getting back to me. This is not really helpful, as I wish to also use SvelteKit's native API endpoints if possible. No solution was found at this stage. Ideally, two different Serverless functions would be created: one for SvelteKit, and one for the golang API endpoints (or one per endpoint, not sure of the internals). Then, the routing would happen first by checking the root API routes (golang), and in case of 404, redirect to the SvelteKit function. Having a way to configure this in the `vercel.json` would be amazing!

**Reply by amyegan** [reply]
Because the root level `/api` currently takes over that route entirely, it's an "all or nothing" option. I think the best solution for that would be a separate project for you golang API. You could even use a [rewrite](https://vercel.com/docs/project-configuration#project-configuration/rewrites) in your main app to make the base URL look the same.

You made a good suggestion to check for matching routes in both locations. I'll mark it as a feature request 😃
